src/storage/src/sink/kafka.rs

```diff
@@ -89,1152 +138,504 @@ where
+    /// The number of messages staged for the currently open transactions. It is reset to zero
+    /// every time a transaction commits.
+    staged_messages: u64,
+    /// The total number bytes staged for the currently open transactions. It is reset to zero
+    /// every time a transaction commits.
+    staged_bytes: u64,
 }
 
-impl KafkaTxProducer {
-    async fn new<'a>(
-        KafkaTxProducerConfig {
-            name,
-            connection,
-            storage_configuration,
-            producer_context,
-            sink_id,
-        }: KafkaTxProducerConfig<'a>,
-    ) -> Result<KafkaTxProducer, ContextCreationError> {
+impl TransactionalProducer {
+    /// Initializes a transcational producer for the sink identified by `sink_id`. After this call
+    /// returns it is guranteed that all previous `TransactionalProducer` instances for the same
+    /// sink have been fenced out (i.e `init_transations()` has been called successfully).
+    async fn new(
+        sink_id: GlobalId,
+        connection: &KafkaSinkConnection,
+        storage_configuration: &StorageConfiguration,
+        statistics: StorageStatistics<SinkStatisticsUpdate, SinkStatisticsMetrics>,
+    ) -> Result<Self, ContextCreationError> {
         let client_id = connection.client_id(&storage_configuration.connection_context, sink_id);
         let transactional_id =
             connection.transactional_id(&storage_configuration.connection_context, sink_id);
 
-        let producer = connection
+        let mut options = BTreeMap::new();
+        // Ensure that messages are sinked in order and without duplicates. Note that this only
+        // applies to a single instance of a producer - in the case of restarts, all bets are off
+        // and full exactly once support is required.
+        options.insert("enable.idempotence", "true".into());
+        // Use the compression type requested by the user.
+        options.insert(
+            "compression.type",
+            connection.compression_type.to_librdkafka_option().into(),
+        );
+        // Increase limits for the Kafka producer's internal buffering of messages. Currently we
+        // don't have a great backpressure mechanism to tell indexes or views to slow down, so the
+        // only thing we can do with a message that we can't immediately send is to put it in a
+        // buffer and there's no point having buffers within the dataflow layer and Kafka. If the
+        // sink starts falling behind and the buffers start consuming too much memory the best
+        // thing to do is to drop the sink. Sets the buffer size to be 16 GB (note that this
+        // setting is in KB)
+        options.insert("queue.buffering.max.kbytes", format!("{}", 16 << 20));
+        // Set the max messages buffered by the producer at any time to 10MM which is the maximum
+        // allowed value.
+        options.insert("queue.buffering.max.messages", format!("{}", 10_000_000));
```

Comment:
Perhaps we want to punt this to a follow up, but I don't think this rationale holds any more! We _do_ have a great way to buffer the input to a sink now: materialized views! The default librdkafka buffer size is probably fine.. if not too large.

Relevant issue: https://github.com/MaterializeInc/materialize/issues/16266

Reply:
definitely worth a todo and followup to validate this!